




	
	
		

	
	
		

	
	
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		
		
		

	
	
		

	
	
		
		
		

	
	
		

	
	
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		
		

	
	
		

	
	
		

	
	
		
		
		

	
	
		
		

	
	
		

	
	
		
		
		

	
	
		

	
	
		
		
		

	
	
		

	
	
		
		
		

	
	
		
		

	
	
		
		
		

	
	
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		
		
		

	
	
		

	
	
		
		
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		
		
		

	
	
		

	
	
		
		
		

	
	
		

	
	
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		
		

	
	
		
		
		

	
	
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		
		

	
	
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		

	
	
		

	
	
		
		
		
		

	
	
		
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		
		

	
	
		

	
	
		
		

	
	
		
		

	
	
		
		

	
	
		
		
		

	
	
		

	
	
		
		
<!DOCTYPE html>
<html lang="en" dir="ltr">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="height=device-height, width=device-width, initial-scale=1.0, minimum-scale=1.0">
    <meta name="generator" content="Hugo 0.117.0">
    <meta name="generator" content="Relearn 5.18.0">
    <meta name="robots" content="noindex, nofollow, noarchive, noimageindex">
    <meta name="description" content="CC 210 Textbook">
    <meta name="author" content="Russell Feldhausen">
    <title>Python Exceptions :: CC 210 Textbook</title>
    <!-- https://github.com/filamentgroup/loadCSS/blob/master/README.md#how-to-use -->
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/fontawesome-all.min.css?1701360991" rel="stylesheet" media="print" onload="this.media='all';this.onload=null;"><noscript><link href="https://ksu-cs-textbooks.github.io/cc210/css/fontawesome-all.min.css?1701360991" rel="stylesheet"></noscript>
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/nucleus.css?1701360991" rel="stylesheet">
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/auto-complete.css?1701360991" rel="stylesheet" media="print" onload="this.media='all';this.onload=null;"><noscript><link href="https://ksu-cs-textbooks.github.io/cc210/css/auto-complete.css?1701360991" rel="stylesheet"></noscript>
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/perfect-scrollbar.min.css?1701360991" rel="stylesheet">
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/fonts.css?1701360991" rel="stylesheet" media="print" onload="this.media='all';this.onload=null;"><noscript><link href="https://ksu-cs-textbooks.github.io/cc210/css/fonts.css?1701360991" rel="stylesheet"></noscript>
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/theme.css?1701360991" rel="stylesheet">
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/theme-auto.css?1701360991" rel="stylesheet" id="variant-style">
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/variant.css?1701360991" rel="stylesheet">
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/print.css?1701360991" rel="stylesheet" media="print">
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/format-print.css?1701360991" rel="stylesheet">
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/ie.css?1701360991" rel="stylesheet">
    <script src="https://ksu-cs-textbooks.github.io/cc210/js/url.js?1701360991"></script>
    <script src="https://ksu-cs-textbooks.github.io/cc210/js/variant.js?1701360991"></script>
    <script>
      // hack to let hugo tell us how to get to the root when using relativeURLs, it needs to be called *url= for it to do its magic:
      // https://github.com/gohugoio/hugo/blob/145b3fcce35fbac25c7033c91c1b7ae6d1179da8/transform/urlreplacers/absurlreplacer.go#L72
      window.index_js_url="https://ksu-cs-textbooks.github.io/cc210/index.search.js";
      var root_url="https://ksu-cs-textbooks.github.io/cc210/";
      var baseUri=root_url.replace(/\/$/, '');
      // translations
      window.T_Copy_to_clipboard = 'Copy to clipboard';
      window.T_Copied_to_clipboard = 'Copied to clipboard!';
      window.T_Copy_link_to_clipboard = 'Copy link to clipboard';
      window.T_Link_copied_to_clipboard = 'Copied link to clipboard!';
      window.T_No_results_found = 'No results found for \u0022{0}\u0022';
      window.T_N_results_found = '{1} results found for \u0022{0}\u0022';
      // some further base stuff
      var baseUriFull='https:\/\/ksu-cs-textbooks.github.io\/cc210/';
      window.variants && variants.init( [ 'auto', 'light-theme', 'dark-theme' ] );
    </script>
    
    <link href="https://ksu-cs-textbooks.github.io/cc210/css/custom.css?1701360991" rel="stylesheet">

  </head>
  <body class="mobile-support print disableInlineCopyToClipboard" data-url="https://ksu-cs-textbooks.github.io/cc210/10-exceptions/06-python/index.html">
    <div id="body" class="default-animation">
      <div id="sidebar-overlay"></div>
      <div id="toc-overlay"></div>
      <nav id="topbar" class="highlightable">
        <div>
          <div id="top-tele-link">
            <a class="print-link" title='Teleprompter View' href="https://ksu-cs-textbooks.github.io/cc210/10-exceptions/06-python/tele.html">
              <i class="fas fa-tv fa-fw"></i>
            </a>
          </div>
          <div id="top-embed-link">
            <a class="print-link" title='Embeddable Version' href="https://ksu-cs-textbooks.github.io/cc210/10-exceptions/06-python/embed.html">
              <i class="fas fa-expand-arrows-alt fa-fw"></i>
            </a>
          </div>
          <div id="breadcrumbs">
            <span id="sidebar-toggle-span">
              <a href="#" id="sidebar-toggle" class="topbar-link" title='Menu (CTRL+ALT+n)'><i class="fas fa-bars fa-fw"></i></a>
            </span>
            <ol class="links" itemscope itemtype="http://schema.org/BreadcrumbList">
              <li itemscope itemtype="https://schema.org/ListItem" itemprop="itemListElement"><a itemprop="item" href="https://ksu-cs-textbooks.github.io/cc210/index.html"><span itemprop="name">CC 210 Textbook</span></a><meta itemprop="position" content="1"> &gt; </li>
              <li itemscope itemtype="https://schema.org/ListItem" itemprop="itemListElement"><a itemprop="item" href="https://ksu-cs-textbooks.github.io/cc210/10-exceptions/index.html"><span itemprop="name">Exceptions</span></a><meta itemprop="position" content="2"> &gt; </li>
              <li itemscope itemtype="https://schema.org/ListItem" itemprop="itemListElement"><span itemprop="name">Python Exceptions</span><meta itemprop="position" content="3"></li>
            </ol>
          </div>
        </div>
      </nav>
      <main id="body-inner" class="highlightable chapter narrow" tabindex="-1">
        <div class="flex-block-wrapper">
          <article class="chapter">
            <header class="headline">
            </header>


<div class="article-subheading">Chapter 6.P</div>
<h1 id="python-exceptions">Python Exceptions</h1>

<p>Exceptions in Python</p>

            <footer class="footline">

            </footer>
          </article>

          <section>
            <h1 class="a11y-only">Subsections of Python Exceptions</h1>
    
    
          <article class="default">
            <header class="headline">
            </header>
<h1 id="common-exceptions">Common Exceptions</h1>

<p>Before we learn about how to detect and handle exceptions in Python, let&rsquo;s review some of the common exceptions and errors we may see in our programs. Each of the headers below is the name of an exception in Python, which is represented by a particular class in the Python programming language.</p>
<h2 id="exception">Exception</h2>
<p>Every exception that can be handled in Python is a subtype of the Exception class. So, when we aren&rsquo;t sure which type of exception to expect, we can always use the Exception class to make sure we catch all of them.</p>
<h2 id="arithmeticerror">ArithmeticError</h2>
<p>An ArithmeticError can occur whenever our program attempts to perform a calculation that would result in an error. This is the base class for various other errors, such as OverflowError and ZeroDivisionError</p>
<h2 id="overflowerror">OverflowError</h2>
<p>An OverflowError occurs when the result of a calculation cannot be represented correctly. Typically this exception is very rare, but can happen when working with older code or libraries.</p>
<h2 id="zerodivisionerror">ZeroDivisionError</h2>
<p>A ZeroDivisionError is thrown whenever the program is asked to either divide by $0$ or perform the modulo operation with $0$ as the second input. Here&rsquo;s an example:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>x <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span> <span style="color:#f92672">/</span> <span style="color:#ae81ff">0</span> <span style="color:#75715e"># throws ZeroDivisionError</span>
</span></span><span style="display:flex;"><span>y <span style="color:#f92672">=</span> <span style="color:#ae81ff">3</span> <span style="color:#f92672">%</span> <span style="color:#ae81ff">0</span> <span style="color:#75715e"># throws ZeroDivisionError</span></span></span></code></pre></div><h2 id="indexerror">IndexError</h2>
<p>An IndexError happens when we try to access an array index that does not exist. Here&rsquo;s a great example:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>array <span style="color:#f92672">=</span> [<span style="color:#ae81ff">1</span>, <span style="color:#ae81ff">2</span>, <span style="color:#ae81ff">3</span>, <span style="color:#ae81ff">4</span>, <span style="color:#ae81ff">5</span>]
</span></span><span style="display:flex;"><span>array[<span style="color:#ae81ff">5</span>] <span style="color:#f92672">=</span> <span style="color:#ae81ff">10</span> <span style="color:#75715e"># throws IndexError</span></span></span></code></pre></div><p>In this example, we are trying to access the 6th element in the array, which is at array index 5. However, since the size of the array is only 5, we&rsquo;ll get an IndexError when we try to execute this code since there is no 6th element.</p>
<h2 id="nameerror">NameError</h2>
<p>In Python, a NameError occurs when we try to use the name of a variable, class, or other item which Python can&rsquo;t find. For example:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>x <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span> <span style="color:#f92672">+</span> y <span style="color:#75715e"># throws NameError</span></span></span></code></pre></div><p>In this code, we are trying to use the variable <code>y</code> without first giving it a value. When we execute this code, it will throw a NameError.</p>
<h2 id="filenotfounderror">FileNotFoundError</h2>
<p>This exception occurs when the program is trying to open a file that does not exist. This is one of the more common errors that programmers must deal with when using files for input, and it is relatively simple to correct. In most cases, we can simply ask the user to provide another file. Here&rsquo;s an example of some code that may cause this exception:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>filename <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;&#34;</span>
</span></span><span style="display:flex;"><span>reader <span style="color:#f92672">=</span> open(filename) <span style="color:#75715e"># throws FileNotFoundError</span></span></span></code></pre></div><p>In this case, we are providing a blank filename, which causes the program to throw a FileNotFoundError.</p>
<h2 id="typeerror">TypeError</h2>
<p>A TypeError occurs when we try to use a variable in a way that is not allowed based on the type of the variable. We&rsquo;ve probably already seen this exception many times when working with string outputs. Here&rsquo;s an example:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>x <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>
</span></span><span style="display:flex;"><span>print(<span style="color:#e6db74">&#34;Number: &#34;</span> <span style="color:#f92672">+</span> x) <span style="color:#75715e"># throws TypeError</span></span></span></code></pre></div><p>In this code, we are using the plus <code>+</code> operator to concatenate two strings together. However, the second variable, <code>x</code>, is an integer, so it is not the correct type for this operation. Therefore, the code will produce a TypeError when executed.</p>
<h2 id="valueerror">ValueError</h2>
<p>A ValueError typically occurs when an input to a method is an acceptable type but has an incorrect value. Here are a couple of examples:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>x <span style="color:#f92672">=</span> int(<span style="color:#e6db74">&#34;12.3&#34;</span>) <span style="color:#75715e"># throws ValueError</span>
</span></span><span style="display:flex;"><span>y <span style="color:#f92672">=</span> float(<span style="color:#e6db74">&#34;abc&#34;</span>) <span style="color:#75715e"># throws ValueError</span></span></span></code></pre></div><p>In both of these examples, the <code>int</code> and <code>float</code> methods are provided with the correct type of input, a string. However, neither of those strings can be converted to the corresponding type, so a ValueError is thrown.</p>
<h2 id="assertionerror">AssertionError</h2>
<p><a href="https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement" target="_blank">Python Assertions</a>
</p>
<p>An AssertionError happens when our code reaches an <em>assertion</em> that fails. An assertion is a check added by the programmer to verify that a particular situation doesn&rsquo;t occur in the code, such as a variable being negative or an array being too large. Python supports the use of a special keyword <code>assert</code> that can be used to add these assertions to our code. By default, the Python interpreter enforces all <code>assert</code> statements, but they can easily be disabled by adding <code>-O</code> as an argument to the <code>python3</code> command when running a program. This is a helpful step when debugging a new application. We&rsquo;ll learn more about assertions and how they can be used effectively in a later chapter.</p>
<p>For now, here&rsquo;s a quick example of an assertion that would produce an AssertionError:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>x <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>
</span></span><span style="display:flex;"><span>y <span style="color:#f92672">=</span> <span style="color:#ae81ff">3</span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">assert</span> x <span style="color:#f92672">+</span> y <span style="color:#f92672">&lt;</span> <span style="color:#ae81ff">7</span> <span style="color:#75715e"># throws AssertionError if enabled</span></span></span></code></pre></div><h2 id="errors">Errors</h2>
<p>Beyond exceptions, there are a few unrecoverable errors that may occur in our Python code. Recall that errors are special types of exceptions that should not be handled by our programs, and should instead be allowed to cause our programs to crash. Some examples are:</p>
<ul>
<li>SyntaxError - occurs when the Python interpreter finds code with invalid syntax</li>
<li>IndentationError - occurs when the Python interpreter  finds improperly indented code</li>
<li>SystemError - occurs when the Python interpreter encounters an internal error</li>
</ul>
<p>Since Python is an interpreted language, these errors will occur when code is running, sometimes even after most of the program is executed. In general, we should not try to catch or correct these errors in our code. Instead, we should allow our program to crash, knowing that we&rsquo;ll have to fix the underlying source code to solve the problem.</p>
<h3 id="references">References:</h3>
<ul>
<li><a href="https://docs.python.org/3/library/exceptions.html" target="_blank">Built-in Exceptions</a>
</li>
<li><a href="https://docs.python.org/3/tutorial/errors.html" target="_blank">Tutorial: Errors and Exceptions</a>
</li>
</ul>

            <footer class="footline">

            </footer>
          </article>

    
    
          <article class="default">
            <header class="headline">
            </header>
<h1 id="exceptions-vs-errors">Exceptions vs. Errors</h1>

<p>Since Python is an interpreted language, it can be hard to tell the difference between <em>exceptions</em> and <em>errors</em>. So, let&rsquo;s take a minute to review those two terms and what they mean.</p>
<h2 id="exceptions">Exceptions</h2>
<p>An <em>exception</em> is any issue encountered while executing a program that can be resolved within the program itself. Typically most exceptions come from incorrect user input, but they can be caused by other issues as well. However, just because an exception can be resolved by the program does not automatically mean that the program does, or must, include a handler for that exception. It just means that it could handle the exception, if the developer included the appropriate handler and code.</p>
<h2 id="errors">Errors</h2>
<p>As discussed before, <em>errors</em> are a special type of exception that cannot be easily handled by the program. These are usually due to issues in the underlying operating system or problems in the code itself. The Python interpreter may return errors such as SyntaxError or IndentationError relating to issues with the code, but it also may return errors such as SystemError or OSError if it encounters a problem that it cannot easily resolve.</p>

            <footer class="footline">

            </footer>
          </article>

    
    
          <article class="default">
            <header class="headline">
            </header>
<h1 id="try-except">Try-Except</h1>

<p><a href="https://www.youtube.com/watch?v=VRg7hkkY4iI">YouTube Video</a></p>
<!-- TODO Update Video? -->
<p><a href="https://ksu-cs-textbooks.github.io/cc210/10-exceptions/06-python/03-try-except/video/">Video Materials</a>
</p>
<p>Programs written in Python can have many exceptions occur on a regular basis. Thankfully, as we learned earlier, it is possible to detect and handle these exceptions in our code directly, in order to prevent our program crashing and causing the user more stress. Let&rsquo;s see how we can perform this step in Python.</p>
<p>Consider the following program, if provided a command-line argument, it treats that string as a file-name and opens it for reading.  If there is no command-lie argument, it defaults to keyboard entry.  It then attempts to get an input and convert that to an integer.</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">import</span> sys
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">class</span> <span style="color:#a6e22e">Try</span>:
</span></span><span style="display:flex;"><span>    
</span></span><span style="display:flex;"><span>    <span style="color:#a6e22e">@classmethod</span>
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">def</span> <span style="color:#a6e22e">main</span>(cls, args):
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">if</span> len(args) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>            reader <span style="color:#f92672">=</span> open(args[<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>            reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span>        x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>        print(x)
</span></span><span style="display:flex;"><span>           
</span></span><span style="display:flex;"><span>        
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">if</span> __name__ <span style="color:#f92672">==</span> <span style="color:#e6db74">&#34;__main__&#34;</span>:
</span></span><span style="display:flex;"><span>    Try<span style="color:#f92672">.</span>main(sys<span style="color:#f92672">.</span>argv)</span></span></code></pre></div><h2 id="try-except">Try-Except</h2>
<p>In Python, we use a <strong>Try-Except</strong> construct to detect and handle exceptions in our code. Let&rsquo;s look at a quick example, and then we can discuss how it works.</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">if</span> len(args) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>    reader <span style="color:#f92672">=</span> open(args[<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>    reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span>  x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>  print(x)
</span></span><span style="display:flex;"><span>    
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">Exception</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error!&#34;</span>)</span></span></code></pre></div><p>First, we use the <code>try</code> keyword, followed by a colon <code>:</code> to begin a block of code. Below that, we indent any lines to be included in the <code>try</code> block. If any exceptions occur in the code indented below a <code>try</code> keyword, we can add <code>except</code> statements to handle that exception.</p>
<p>Directly following the <code>try</code> block, we must include at least one <code>except</code> statement. The <code>except</code> keyword is followed by the name of an exception that we&rsquo;d like to catch. Optionally we can add the <code>as</code> keyword and the name of a variable to represent that exception. Finally, we include a colon <code>:</code> and indent a block of code below to act as the exception handler.</p>
<p>In this example, we are just catching the generic <code>Exception</code> type, which will match any catchable exception. We&rsquo;ll see how we can detect specific exceptions in a later example. We can use the variable for the exception, in this case <code>e</code>, to access additional details about the exception we&rsquo;ve detected.</p>
<p style="padding:10"><b>Note:</b> nearly every other language uses the syntax "try catch" instead of "try except". </p>
<h2 id="which-exceptions-to-catch">Which Exceptions to Catch?</h2>
<p>When writing code, it can sometimes be very difficult to even know which exceptions to expect from a particular piece of code. Thankfully, the <a href="https://docs.python.org/3/library/exceptions.html" target="_blank">Python Documentation</a>
 includes quite a bit of information about the possible exceptions that could occur in a program.</p>
<p>Let&rsquo;s return to our earlier example. Here is the code contained in the <code>try</code> block:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">if</span> len(args) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>  reader <span style="color:#f92672">=</span> open(args[<span style="color:#ae81ff">1</span>])  
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>  reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>print(x)</span></span></code></pre></div><p>While this may look like a very simple few lines of code, there are actually several exceptions that could be produced here:</p>
<ul>
<li>FileNotFoundError - if the file in <code>args[1]</code> is not valid</li>
<li>NameError - if we forget to include <code>import sys</code> at the top of the file</li>
<li>ValueError - if the input cannot be converted to an integer</li>
<li>IndexError - <code>args[1]</code> could produce this exception, but we&rsquo;ve already checked that the length is greater than 1, so it won&rsquo;t happen in practice.</li>
</ul>
<p>When writing truly bulletproof code, it is a good idea to attempt to catch and handle all of these exceptions if possible. You can always refer to the official Python documentation for a list of exceptions that could occur. Unfortunately, Python does not do a great job of documenting which exceptions could be thrown by specific methods compared to Java. Later on in this chapter we&rsquo;ll discuss some best practices when it comes to detecting and handling exceptions in code.</p>
<h2 id="handling-multiple-exceptions">Handling Multiple Exceptions</h2>
<p>Of course, we can add multiple <code>except</code> statements after any <code>try</code> block to catch different types of exceptions.</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># If an argument is present, we are reading from a file</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># specified in args[1]</span>
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">if</span> len(args) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>    reader <span style="color:#f92672">=</span> open(args[<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># If no argument, read from stdin  </span>
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>    reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>  x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>  print(x)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">FileNotFoundError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error: File Not Found!&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error: Input Does Not Match Expected Format!&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">OSError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error: OS Exception!&#34;</span>)</span></span></code></pre></div><p>In the example code above, we see three different <code>except</code> statements, each of which will handle a different type of exception. When an exception occurs in the code contained in a <code>try</code> block, the Python interpreter will create the exception, and then it will search for the first handler that matches. So, it will begin with the first <code>except</code> statement, and see if the exception created matches the type of the exception in parenthesis. If so, it will execute the code inside of that <code>except</code> block. If not, it will continue to the next <code>except</code> statement. If none of the <code>except</code> statements match the exception, then it will be <em>thrown</em> and cause the program to stop executing.</p>
<h2 id="exception-hierarchy">Exception Hierarchy</h2>
<p>The exceptions in Python form a hierarchical structure, meaning that an exception may match multiple types. For example, all exceptions that programs can catch are based on the generic Exception type. So, FileNotFoundError and IndexError would both match the Exception type.</p>
<p>In addition, many exceptions are descended from the OSError type. For example, FileNotFoundError is based on OSError, which itself is based on the Exception type. So, a FileNotFoundError would match any of those three types of exceptions. It can make things a bit tricky!</p>
<p>So, how can we know what types of exceptions we are dealing with, and what the hierarchy is? Thankfully, the <a href="https://docs.python.org/3/library/exceptions.html#exception-hierarchy" target="_blank">Python Documentation</a>
 contains information about most possible exceptions, including the entire hierarchy of those exceptions.</p>
<p>For example, here is a screenshot from the <a href="https://docs.python.org/3/library/exceptions.html#exception-hierarchy" target="_blank">Python Documentation</a>
 page showing the hierarchy of OSError:</p>
<p>
<a href="#image-5c19cce7c9835fc5f77ebb8a55d3957b" class="lightbox-link">
<img src="https://ksu-cs-textbooks.github.io/cc210/images/10-except/8.7.p.3.filenotfound.png" alt="OSError Hierarchy" style="height: auto; width: auto;" loading="lazy">
</a>
<a href="javascript:history.back();" class="lightbox" id="image-5c19cce7c9835fc5f77ebb8a55d3957b">
<img src="https://ksu-cs-textbooks.github.io/cc210/images/10-except/8.7.p.3.filenotfound.png" alt="OSError Hierarchy" class="lightbox-image" loading="lazy">
</a></p>
<p>Because of this, we must be careful about how we order the catch statements. In general, we want to place the more specific exceptions first (the ones further down the hierarchy), and the more generic exceptions later.</p>
<p>Let&rsquo;s look at an example of poor ordering of exception handlers:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># If an argument is present, we are reading from a file</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># specified in args[1]</span>
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">if</span> len(args) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>    reader <span style="color:#f92672">=</span> open(args[<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># If no argument, read from stdin  </span>
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>    reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>  x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>  print(x)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">OSError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error: OS Exception!&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">FileNotFoundError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error: File Not Found!&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error: Input Does Not Match Expected Format!&#34;</span>)</span></span></code></pre></div><p>In the code above, we are catching the OSError first. So, if the code produces a FileNotFoundError, it will be caught and handled by the first <code>except</code> statement, since FileNotFoundError is also an OSError. Therefore, we wouldn&rsquo;t want to order our <code>except</code> statements in this way.</p>
<h2 id="try-it">Try It!</h2>
<p>Let&rsquo;s see if we can write a very simple program to catch and handle a few common exceptions. Here&rsquo;s some starter code for <code>main</code>&rsquo;s:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">if</span> len(args) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>  reader <span style="color:#f92672">=</span> open(args[<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>  reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>y <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>z <span style="color:#f92672">=</span> x <span style="color:#f92672">/</span> y
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>print(z)</span></span></code></pre></div><p>For this example, place this starter code in <code>Try.py</code>, open to the left, and modify it to catch and handle the following exceptions by printing the given error messages:</p>
<ul>
<li>FileNotFoundError - print &ldquo;Error: File Not Found!&rdquo;</li>
<li>ValueError - print &ldquo;Error: Input Does Not Match Expected Format!&rdquo;</li>
<li>ZeroDivisionError - print &ldquo;Error: Divide by Zero!&rdquo;</li>
</ul>
<p>Any other exceptions can be ignored.</p>

  
    
    
  
  
<div class="box notices cstyle noiframe">
  <div class="box-label"><i class="fa-fw fas fa-times-circle"></i> Web Only</div>
  <div class="box-content">

<p>This content is presented in the course directly through Codio. Any references to interactive portions are only relevant for that interface. This content is included here as reference only.</p>
</div>
</div>

            <footer class="footline">

            </footer>
          </article>

          <section>
            <h1 class="a11y-only">Subsections of Try-Except</h1>
    
    
          </section>
    
    
          <article class="default">
            <header class="headline">
            </header>
<h1 id="raise">Raise</h1>

<p><a href="https://www.youtube.com/watch?v=WTo7Xdrg0Ec">YouTube Video</a></p>
<p><a href="https://ksu-cs-textbooks.github.io/cc210/10-exceptions/06-python/04-raise/video/">Video Materials</a>
</p>
<p>Beyond catching exceptions, there are a few other important concepts to learn related to exceptions. Let&rsquo;s go over a couple of them.</p>
<h2 id="raise">Raise</h2>
<p>We can generate our own exceptions using the <code>raise</code> keyword. This allows us to create new exceptions whenever needed. Let&rsquo;s look at an example:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">import</span> sys
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">class</span> <span style="color:#a6e22e">Raise</span>:
</span></span><span style="display:flex;"><span>    
</span></span><span style="display:flex;"><span>    <span style="color:#a6e22e">@classmethod</span>
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">def</span> <span style="color:#a6e22e">main</span>(cls, args):       
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">if</span> len(args) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>          reader <span style="color:#f92672">=</span> open(args[<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>          reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span>        x <span style="color:#f92672">=</span> float(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>        y <span style="color:#f92672">=</span> float(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">if</span> y <span style="color:#f92672">==</span> <span style="color:#ae81ff">0</span>:
</span></span><span style="display:flex;"><span>          <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">ValueError</span>(<span style="color:#e6db74">&#34;Cannot divide by zero&#34;</span>)
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>          z <span style="color:#f92672">=</span> x <span style="color:#f92672">/</span> y
</span></span><span style="display:flex;"><span>          print(z)</span></span></code></pre></div><p>In this code, we are asking the user to input two floating point numbers. Then, we will output the first number divided by the second. However, we want to avoid the situation where the second number, the divisor, is $0$. Instead of causing a DivideByZeroError, we can use an <strong>If-Then</strong> statement and the <code>raise</code> keyword to send a ValueError instead.</p>
<p>Following the <code>raise</code> keyword, we must create a new Exception. In this case, we are creating a new ValueError, but any of the exception types we&rsquo;ve learned about so far will work the same way. Inside of the parentheses, we can provide a helpful error message to go along with this exception.</p>
<h2 id="try-it">Try It!</h2>
<p>Let&rsquo;s see if we can use these keywords in another example. We&rsquo;ll start with this code in <code>main()</code>:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">if</span> len(args) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>  reader <span style="color:#f92672">=</span> open(args[<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>  reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin</span></span></code></pre></div><p>Place this code in <code>Raise.py</code> and modify it to do the following:</p>
<ol>
<li>If the user provides an invalid file name in the arguments array, it should throw a FileNotFoundError.</li>
<li>It should read a single string of input from the user. To make this simple, don&rsquo;t forget to use the <code>strip()</code> method to remove any extra whitespace and trailing newline characters!</li>
<li>If the string does not begin with a capital letter, it should throw a new ValueError that contains the error message &ldquo;Proper Capitalization Required!&rdquo;</li>
<li>Of course, if the string is empty, it should throw an IndexError with the message &ldquo;String is Empty!&rdquo; when it tries to access the first character of the string.</li>
<li>Otherwise, it should print the string to the terminal.</li>
</ol>

  
    
    
  
  
<div class="box notices cstyle noiframe">
  <div class="box-label"><i class="fa-fw fas fa-times-circle"></i> Web Only</div>
  <div class="box-content">

<p>This content is presented in the course directly through Codio. Any references to interactive portions are only relevant for that interface. This content is included here as reference only.</p>
</div>
</div>
<h3 id="passing-back-text">Passing Back Text</h3>
<p>The construct <code>raise &lt;exception&gt;(&lt;string&gt;)</code> returns an <code>&lt;class:ValueError&gt;</code> object.  Like all objects it has a <code>__str__</code> method that is called when <code>print</code>ed or used with <code>str()</code>.  The value returned is by <code>str()</code> is <code>&lt;string&gt;</code>.</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">ValueError</span>(<span style="color:#e6db74">&#34;demo&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>   print(type(e)) <span style="color:#75715e"># &lt;class &#39;ValueError&#39;&gt;</span>
</span></span><span style="display:flex;"><span>   print(e)       <span style="color:#75715e"># &#34;demo&#34;</span></span></span></code></pre></div><p>This behavior is convenient when a single type of exception can be triggered by different causes.  Each <code>raise ..</code>  can have a different explanatory string or error message.</p>

            <footer class="footline">

            </footer>
          </article>

          <section>
            <h1 class="a11y-only">Subsections of Raise</h1>
    
    
          </section>
    
    
          <article class="default">
            <header class="headline">
            </header>
<h1 id="else--finally">Else &amp; Finally</h1>

<p><a href="https://www.youtube.com/watch?v=L7hep4SmBpo">YouTube Video</a></p>
<p><a href="https://ksu-cs-textbooks.github.io/cc210/10-exceptions/06-python/05-else-finally/video/">Video Materials</a>
</p>
<p>When dealing with exceptions in our code, sometimes we have an operation that must be executed, even if the earlier code throws an exception. In that case, we can use the <code>finally</code> keyword to make sure that the correct code is executed. In addition, we may have an operation that should only happen if there are no exceptions in the previous code. For that, we can use the <code>else</code> keyword.</p>
<h2 id="finally">Finally</h2>
<p>To understand how the <code>finally</code> keyword works, let&rsquo;s take a look at an example:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">import</span> sys
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>  x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">if</span> x <span style="color:#f92672">&lt;</span> <span style="color:#ae81ff">0</span>:
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">ValueError</span>(<span style="color:#e6db74">&#34;Input must be greater than 0!&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Value Error: </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">&#34;</span><span style="color:#f92672">.</span>format(e))
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">finally</span>:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Finally Block&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>print(<span style="color:#e6db74">&#34;After Try&#34;</span>)</span></span></code></pre></div><p>This program will read an integer from the terminal. If the integer is greater than or equal to 0, it will do nothing except print the &ldquo;Finally Block&rdquo; and &ldquo;After Try&rdquo; messages. However, if the input is less than 0, it will throw a ValueError. Finally, if the input is not a number, then a ValueError will also be thrown and handled. In each of those cases, it will also print the &ldquo;Finally Block&rdquo; message.</p>
<p>Let&rsquo;s run this program a few times and see how it works. First, we&rsquo;ll provide the input &ldquo;5&rdquo;:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-tex" data-lang="tex"><span style="display:flex;"><span><span style="color:#e6db74">$</span> python<span style="color:#ae81ff">3</span> <span style="color:#ae81ff">8</span>p<span style="color:#f92672">-</span>except<span style="color:#f92672">/</span>Finally.py
</span></span><span style="display:flex;"><span><span style="color:#ae81ff">5</span>
</span></span><span style="display:flex;"><span>Finally Block
</span></span><span style="display:flex;"><span>After Try</span></span></code></pre></div><p>Here, we can see that the code in the <code>finally</code> block always runs when the program is finished executing the statements in the <code>try</code> block.</p>
<p>Let&rsquo;s run it again, this time with &ldquo;-5&rdquo; as the input:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-tex" data-lang="tex"><span style="display:flex;"><span><span style="color:#e6db74">$</span> python<span style="color:#ae81ff">3</span> <span style="color:#ae81ff">8</span>p<span style="color:#f92672">-</span>except<span style="color:#f92672">/</span>Finally.py
</span></span><span style="display:flex;"><span><span style="color:#f92672">-</span><span style="color:#ae81ff">5</span>
</span></span><span style="display:flex;"><span>Value Error: Input must be greater than <span style="color:#ae81ff">0</span><span style="color:#f92672">!</span>
</span></span><span style="display:flex;"><span>Finally Block
</span></span><span style="display:flex;"><span>After Try</span></span></code></pre></div><p>Here, we can see that it prints the error message caused by the ValueError, then proceeds to the <code>finally</code> block. So, even if an exception is thrown while inside of the <code>try</code> block, the code in the <code>finally</code> block is <strong>always</strong> executed once the <code>try</code> block and any exception handlers are finished.</p>
<p>Here&rsquo;s one more example, this time with &ldquo;abc&rdquo; as the input:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-tex" data-lang="tex"><span style="display:flex;"><span><span style="color:#e6db74">$</span> python<span style="color:#ae81ff">3</span> <span style="color:#ae81ff">8</span>p<span style="color:#f92672">-</span>except<span style="color:#f92672">/</span>Finally.py
</span></span><span style="display:flex;"><span>abc
</span></span><span style="display:flex;"><span>Value Error: invalid literal for int<span style="color:#f92672">()</span> with base <span style="color:#ae81ff">10</span>: &#39;abc\n&#39;
</span></span><span style="display:flex;"><span>Finally Block
</span></span><span style="display:flex;"><span>After Try</span></span></code></pre></div><p>Once again, we see that the code first handles the ValueError, and then it will execute the code in the <code>finally</code> block.</p>
<p>In a later chapter, we&rsquo;ll learn more about how to use the <code>finally</code> block to perform important tasks such as closing open files and making sure we don&rsquo;t leave the system in an unstable state when we handle an exception.</p>
<h2 id="else">Else</h2>
<p>Python also allows the <code>else</code> keyword to be used with a <strong>Try-Except</strong> statement. It works similarly to the way Python handles the <code>else</code> keyword with loops. Let&rsquo;s look at an example:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">import</span> sys
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>  x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">if</span> x <span style="color:#f92672">&lt;</span> <span style="color:#ae81ff">0</span>:
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">ValueError</span>(<span style="color:#e6db74">&#34;Input must be greater than 0!&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Value Error: </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">&#34;</span><span style="color:#f92672">.</span>format(e))
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;No Errors!&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">finally</span>:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Finally Block&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>print(<span style="color:#e6db74">&#34;After Try&#34;</span>)</span></span></code></pre></div><p>The code above is the same example used to demonstrate the <code>finally</code> keyword, but now there is an <code>else</code> block added. The code in the <code>else</code> block will be executed only if the entire code in the <code>try</code> block is able to execute without any errors or exceptions being raised.</p>
<p>So, if we run this program with the input <code>5</code>, we&rsquo;ll get the following:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-tex" data-lang="tex"><span style="display:flex;"><span><span style="color:#e6db74">$</span> python<span style="color:#ae81ff">3</span> <span style="color:#ae81ff">8</span>p<span style="color:#f92672">-</span>except<span style="color:#f92672">/</span>Finally.py
</span></span><span style="display:flex;"><span><span style="color:#ae81ff">5</span>
</span></span><span style="display:flex;"><span>No Errors<span style="color:#f92672">!</span>
</span></span><span style="display:flex;"><span>Finally Block
</span></span><span style="display:flex;"><span>After Try</span></span></code></pre></div><p>Since this program does not encounter any exceptions, the <code>else</code> block is executed immediately after the <code>try</code> block, but before the <code>finally</code> block. However, if we provide an invalid input, such as <code>-5</code>, the output will look like this instead:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-tex" data-lang="tex"><span style="display:flex;"><span><span style="color:#e6db74">$</span> python<span style="color:#ae81ff">3</span> <span style="color:#ae81ff">8</span>p<span style="color:#f92672">-</span>except<span style="color:#f92672">/</span>Finally.py
</span></span><span style="display:flex;"><span><span style="color:#f92672">-</span><span style="color:#ae81ff">5</span>
</span></span><span style="display:flex;"><span>Value Error: Input must be greater than <span style="color:#ae81ff">0</span><span style="color:#f92672">!</span>
</span></span><span style="display:flex;"><span>Finally Block
</span></span><span style="display:flex;"><span>After Try</span></span></code></pre></div><p>Since this program encountered an exception, the code in the <code>else</code> block is not executed. However, the code in the <code>finally</code> block is still executed, since it will happen regardless of the outcome of the <code>try</code> block.</p>
<p>Finally, it is possible to include an <code>else</code> block without a <code>finally</code> block if desired. The only rule is that the <code>else</code> block must come before the <code>finally</code> block, and both of them should be after any <code>except</code> blocks to handle exceptions.</p>

            <footer class="footline">

            </footer>
          </article>

          <section>
            <h1 class="a11y-only">Subsections of Else &amp; Finally</h1>
    
    
          </section>
    
    
          <article class="default">
            <header class="headline">
            </header>
<h1 id="with">With</h1>

<!--
<a href="https://www.youtube.com/watch?v=Mr6OPmyKRCM">YouTube Video</a>
-->
<!--[Video Materials](/10-exceptions/06-python/06-with/video/)-->
<!-- TODO Redo Video so it doesn't close stdin -->
<p>Lastly, Python includes another new statement that is relevant to dealing with exceptions, called the <strong>With</strong> statement.</p>
<h2 id="with">With</h2>
<p>To understand how the <strong>With</strong> statement works in Python, let&rsquo;s look at an example:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">import</span> sys
</span></span><span style="display:flex;"><span><span style="color:#f92672">import</span> traceback
</span></span><span style="display:flex;"><span><span style="color:#f92672">import</span> io
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">if</span> len(sys<span style="color:#f92672">.</span>argv) <span style="color:#f92672">&gt;</span><span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">with</span> open(sys<span style="color:#f92672">.</span>argv[<span style="color:#ae81ff">1</span>]) <span style="color:#66d9ef">as</span> input_file:
</span></span><span style="display:flex;"><span>            data <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;&#34;</span><span style="color:#f92672">.</span>join(input_file<span style="color:#f92672">.</span>readlines())
</span></span><span style="display:flex;"><span>            reader <span style="color:#f92672">=</span> io<span style="color:#f92672">.</span>StringIO(data)
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>        reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin 
</span></span><span style="display:flex;"><span>    
</span></span><span style="display:flex;"><span>    x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>    print(x)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;ValueError: </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">&#34;</span><span style="color:#f92672">.</span>format(e))
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">FileNotFoundError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;FileNotFoundError: </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">&#34;</span><span style="color:#f92672">.</span>format(e))</span></span></code></pre></div><p>In this example, we are first using the standard <strong>If</strong> statement to either open a file provided as the first command-line argument, or else we are reading input from the terminal.</p>
<p>If there is a file provided, we open it using a <strong>with</strong> statement.  This statement is used to create a <em>context manager</em> in Python. In short, it allows languages developers to define things that must always be done before and after a resource, such as an open file, is used. In this example, in addition to opening the file, it will make sure the <code>infile</code> is closed when we leave the with block.  The other statements manipulate the file into StringIO object that uses the same methods as sys.stdin.</p>
<p>When the code in the <code>with</code> statement throws an exception, Python will automatically perform the cleanup operations for the resource listed in the <code>with</code>. So, when we are reading input from a file, it would close that file and make sure that it isn&rsquo;t damaged or left open when our program crashes.</p>
<p>In addition, any additional exceptions thrown when trying to close the file are <em>suppressed</em> by the system, so we only see the exception that caused the initial error. This is much better than using a <code>finally</code> statement to close the file, since we&rsquo;d have to run the risk of throwing a second exception inside of the <code>finally</code> statement.</p>
<p>This type of statement is great when writing programs that will handle large amounts of input, either by connecting to a database, reading from a file, or using the Internet to communicate with a server. Many standard Python libraries can be used in this way, especially if they require a <code>close()</code> or other cleanup method to be called when we are finished using them. So, a <code>with</code> statement is a great way to make sure those programs are able to handle exceptions without leaving the system in an unstable state.</p>
<p style="color:red"><b>WARNING:</b> avoid using **with** in conjunction with stdin. <br> `with sys.stdin as reader:` is bad.  Python will dutifully close the stdin at the end of the with-block -- making it unavailable to the program a second time.  This would be like having to close and reopen Word, because every time you hit `return` the keyboard stopped working.</p>
<p>More on <code>with ... as ...:</code> after we cover File input.</p>

            <footer class="footline">

            </footer>
          </article>

          <section>
            <h1 class="a11y-only">Subsections of With</h1>
    
    
          </section>
    
    
          <article class="default">
            <header class="headline">
            </header>
<h1 id="best-practices">Best Practices</h1>

<p>Now that we&rsquo;ve seen lots of information about how to throw, catch, and handle exceptions in Python, it is a great time to discuss some of the best practices we can apply in our code to make it as readable and bulletproof as possible. While these aren&rsquo;t strict rules that must be followed all the time, they are great things to keep in mind as we write code that deals with exceptions.</p>
<h2 id="leave-no-exception-unhandled">Leave No Exception Unhandled</h2>
<p>As much as possible, we should write our code to handle any exception we can reasonably expect our users to run into when using our programs. We cannot assume that users will always provide the correct input in the correct format, and a single typo by a user should not cause our entire program to crash.</p>
<p>Instead, we should always use <strong>Try-Except</strong> statements whenever possible to detect and handle those errors as simply as possible. In addition, if the program is interactive, we can combine that approach with the use of loops to prompt the user to provide new input to resolve the error.</p>

  
  
<div class="box notices cstyle info">
  <div class="box-label"><i class="fa-fw fas fa-info-circle"></i> Don&rsquo;t Substitute Unchecked Exception Handling for Value Checking</div>
  <div class="box-content">

<p>You can go &ldquo;try except&rdquo; crazy.  Exception handling is powerful, but also slow, really really slow.</p>
<p>You can write something like</p>
<div class="wrap-code highlight"><pre tabindex="0"><code>try:
    ratio = x / y
except ZeroDivisionError as e:
    print(&#34;Cannot divide by zero&#34;)</code></pre></div><p>but</p>
<div class="wrap-code highlight"><pre tabindex="0"><code>if y !=0:
    ratio = x / y;
else:
    print(&#34;Cannot divide by zero&#34;)</code></pre></div><p>executes a lot more efficiently.</p>
<p>Many exceptions can be avoided through value checking.  In this course, we may direct you to throw and catch exceptions of this type for practice in exception coding.</p>
</div>
</div>
<p>The <a href="https://docs.python.org/3/" target="_blank">Python Documentation</a>
 is an okay resource to determine which exceptions could be thrown by any methods used in our code.  However Python does not provide a standard or comprehensive list of the Exceptions and conditions which cause them for all methods in all modules&ndash;nor is their enumeration in the Python Docs in a standard place or format.</p>
<h2 id="dp-not-use-exception-handling-for-control-flow">Dp Not Use Exception Handling for Control Flow</h2>
<p>You might be tempted to do something like this</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>    reader <span style="color:#f92672">=</span> open(sys<span style="color:#f92672">.</span>argv[<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">Exception</span>:
</span></span><span style="display:flex;"><span>    reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin </span></span></code></pre></div><p>Where if there is any problem accessing sys.argv[1] as a file, such as it was not provided or it does not exist, you just use the keyboard.  Here you are letting an exception control what is happening without telling the user there was a problem.   Better would be</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">if</span> len(sys<span style="color:#f92672">.</span>argv) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>        reader<span style="color:#f92672">=</span> open(sys<span style="color:#f92672">.</span>arv[<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">OSError</span>:
</span></span><span style="display:flex;"><span>        print (<span style="color:#e6db74">&#34;there was a problem accessing </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">, please check the file and try again&#34;</span><span style="color:#f92672">.</span>format(sys<span style="color:#f92672">.</span>argv[<span style="color:#ae81ff">1</span>]))
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">return</span> <span style="color:#75715e"># here we assume this will exit the program</span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>    reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin</span></span></code></pre></div><p>Programmers do not expect control flow decisions to be made <b>in</b> exception-handlers.  It makes analysis of the program very difficult.</p>
<h2 id="be-specific">Be Specific</h2>
<p>We should also always strive to use specific exceptions in our <code>except</code> statements whenever we can. So, if we are opening a file, we should include an <code>except</code> statement for a FileNotFoundError, and not just a single generic <code>Exception</code>. Even though this means we may have to include several <code>except</code> statements, it is much better in the long run because it allows us to know exactly what happened when our code has a problem.</p>
<p>In addition, we should always make sure we catch the most specific exceptions first, before any more generic exceptions. As we saw earlier, if we try to catch an OSError before a FileNotFoundError, we won&rsquo;t be able to tell when we&rsquo;ve reached a FileNotFoundError at all, since it is a subtype of the OSError.</p>
<h2 id="use-messages-and-stack-traces">Use Messages and Stack Traces</h2>
<p>The Python Exception class includes some very helpful methods we can use to get additional information when we encounter an exception.</p>
<p>Here&rsquo;s an example to show what we can learn:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">import</span> sys
</span></span><span style="display:flex;"><span><span style="color:#f92672">import</span> traceback
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>  x <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span> <span style="color:#f92672">/</span> <span style="color:#ae81ff">0</span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ZeroDivisionError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error! Trying to Divide by Zero!&#34;</span>)
</span></span><span style="display:flex;"><span>  traceback<span style="color:#f92672">.</span>print_exc(file<span style="color:#f92672">=</span>sys<span style="color:#f92672">.</span>stdout)</span></span></code></pre></div><p>When this code is executed, it will produce the following output:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-tex" data-lang="tex"><span style="display:flex;"><span><span style="color:#e6db74">$</span> python<span style="color:#ae81ff">3</span> <span style="color:#ae81ff">8</span>p<span style="color:#f92672">-</span>except<span style="color:#f92672">/</span>StackTrace.py
</span></span><span style="display:flex;"><span>Error<span style="color:#f92672">!</span> Trying to Divide by Zero<span style="color:#f92672">!</span>
</span></span><span style="display:flex;"><span>Traceback <span style="color:#f92672">(</span>most recent call last<span style="color:#f92672">)</span>:
</span></span><span style="display:flex;"><span>  File &#34;<span style="color:#ae81ff">8</span>p<span style="color:#f92672">-</span>except<span style="color:#f92672">/</span>StackTrace.py&#34;, line <span style="color:#ae81ff">5</span>, in &lt;module&gt;
</span></span><span style="display:flex;"><span>    x <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span> <span style="color:#f92672">/</span> <span style="color:#ae81ff">0</span>
</span></span><span style="display:flex;"><span>ZeroDivisionError: division by zero</span></span></code></pre></div><p>On the first line, we are printing our own error message. Below that, we are using <code>traceback.print_exc(file=sys.stdout)</code> to print a full stack trace showing the location of the error in code. In this example, we included <code>file=sys.stdout</code> because the method will print to <code>sys.stderr</code> by default, a different output stream. This just makes it easier to see all of our output on the terminal for now.</p>
<p>In general, it is best to only show a short error message to users, but in some cases it may be better to add our own message. The text &ldquo;division by zero&rdquo; isn&rsquo;t very clear, even to developers. Instead, we could say &ldquo;Error: divisor cannot be zero&rdquo; to make the error message clearer.</p>
<p>The information contained in the stack trace is very helpful to developers when trying to debug and fix problems in the code, but that information is very confusing to users. As we learn how to build more advanced programs, we&rsquo;ll see how to record and log those error messages and stack traces for debugging, but hide them from our users.</p>
<p>To see what other methods are available, refer to the <a href="https://docs.python.org/3/library/traceback.html" target="_blank">Traceback</a>
 library in the Python documentation.</p>
<h2 id="dont-ignore-exceptions">Don&rsquo;t Ignore Exceptions</h2>
<p>Another big problem is that many developers tend to catch exceptions, only to ignore them so their program doesn&rsquo;t crash. Here&rsquo;s a quick example:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">import</span> sys
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>x <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>
</span></span><span style="display:flex;"><span>reader <span style="color:#f92672">=</span> sys<span style="color:#f92672">.</span>stdin
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>  x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">Exception</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">pass</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># do nothing</span></span></span></code></pre></div><p>In this code, if the user inputs something that can&rsquo;t be converted to an integer, the code just silently ignores the exception and proceeds with <code>x</code> still storing the value $0$. While that may not cause issues, it can also make it very frustrating to debug later issues or changes in this program. So, it is always best to output an error message when an exception occurs, even if it can be easily ignored without additional input from the user. This will make it easier to debug additional issues later on in development.</p>
<h2 id="bulletproof-code">Bulletproof Code</h2>
<p>Here&rsquo;s a simple program that asks the user for input, and will repeat the question until a valid input is received. This code is designed to handle many common situations and exceptions:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">import</span> sys
</span></span><span style="display:flex;"><span><span style="color:#f92672">import</span> traceback
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>x <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">with</span> open(sys<span style="color:#f92672">.</span>argv[<span style="color:#ae81ff">1</span>]) <span style="color:#66d9ef">as</span> reader:
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">while</span> x <span style="color:#f92672">&lt;=</span> <span style="color:#ae81ff">0</span>:
</span></span><span style="display:flex;"><span>      <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>        x <span style="color:#f92672">=</span> int(reader<span style="color:#f92672">.</span>readline())
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">if</span> x <span style="color:#f92672">&lt;=</span> <span style="color:#ae81ff">0</span>:
</span></span><span style="display:flex;"><span>          print(<span style="color:#e6db74">&#34;Error: Negative Integer Detected!&#34;</span>)
</span></span><span style="display:flex;"><span>      <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>        print(<span style="color:#e6db74">&#34;Error: Integer Not Found!&#34;</span>)
</span></span><span style="display:flex;"><span>      <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">Exception</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>        print(<span style="color:#e6db74">&#34;Error: Unknown Input Error&#34;</span>)
</span></span><span style="display:flex;"><span>        sys<span style="color:#f92672">.</span>exit() <span style="color:#75715e"># probably can&#39;t recover, so stop executing</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;the file contained </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">&#34;</span><span style="color:#f92672">.</span>format(x))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">FileNotFoundError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error: Unable to File Open Reader!&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">Exception</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>  print(<span style="color:#e6db74">&#34;Error: Unable to File Open Reader!&#34;</span>)</span></span></code></pre></div><p>There are several items in this code that help make it very bulletproof:</p>
<ol>
<li>First, the program uses a <strong>Try-Except</strong> statement containing a <strong>With</strong> statement to make sure that the file will be properly closed when we are done with it. As we&rsquo;ve discussed, this is a good practice when handling input via files, but really any input object can be handled in this way.</li>
<li>The outer <strong>Try-Except</strong> statement has a two <code>except</code> blocks.  The first handles not finding the file (the most likely exception).  The second captures anything else that might go wrong with the reader .</li>
<li>Inside of the <strong>While</strong> loop, there is a second <strong>Try-Except</strong> statement to handle errors that come from reading an individual input, such as a ValueError.</li>
<li>In addition, that inner <strong>Try-Except</strong> statement will also catch any generic Exceptions that might occur when reading input. In that case, we use the <code>sys.exit()</code> method to stop executing the program, since it may be difficult to continue to read input in that case. Thankfully, since we are using a <strong>With</strong> block, the reader will be correctly closed for us automatically. In addition, any <code>finally</code> blocks would be executed before the program stops.</li>
</ol>

  
  
<div class="box notices cstyle warning">
  <div class="box-label"><i class="fa-fw fas fa-exclamation-triangle"></i> Don&rsquo;t Exit Early!</div>
  <div class="box-content">

<p>We have not discussed the <code>sys.exit()</code> method yet, but you may have come across it while reading code online. In Python, the <code>sys.exit()</code> method is used to exit the Python interpreter safely from within a program</p>
<p>For now, if we find that we need to exit our program when handling an error, we can safely use the <code>sys.exit()</code> method as shown in the example above. It will make sure that any <code>finally</code> blocks and <strong>With</strong> statements are executed properly.</p>
<p>There are some other ways to exit the Python interpreter that you can find online, but they may do so without properly closing any resources or executing <code>finally</code> blocks. They can also make your code much more difficult to maintain and test. This can be very dangerous!</p>
</div>
</div>
<p>Of course, there are many things that can be done differently in this code, depending on our preferences and how we&rsquo;d like our program to function. This is just one possible way to build useful code that handles several exceptions.</p>

            <footer class="footline">

            </footer>
          </article>

    
    
          <article class="default">
            <header class="headline">
            </header>
<h1 id="a-worked-example">A Worked Example</h1>

<!--
<a href="https://www.youtube.com/watch?v=XaDEahMnjgU">YouTube Video</a>
-->
<!--[Video Materials](/10-exceptions/06-python/08-example/video/)-->
<!-- TODO Redo Video -->
<p>Let&rsquo;s work through an entire example program together to see how we can build programs that handle exceptions quickly and easily, without allowing the program to crash when invalid input is received.</p>
<h2 id="problem-statement">Problem Statement</h2>
<p>Consider the following problem statement for a driver-class program <code>Example.py</code>:</p>
<blockquote>
<p>Write a program to accept a list of numbers, one per line from the terminal.</p>
</blockquote>
<blockquote>
<p>The numbers can either be whole numbers or floating point numbers. The program should continue to accept input until a number equal to $0$ is input, or a blank line is received. Of course, if the input in a file ends before a $0$, that can be treated as the end of the input.</p>
</blockquote>
<blockquote>
<p>Once $0$ is received or the input ends, the program should print the largest and smallest number provided as input (not including the $0$ to end input). Sample outputs are shown below.</p>
</blockquote>
<blockquote>
<p>The program should catch and handle all common exceptions. When an exception is caught, the program should print the name of the exception, followed by the message included in the exception. It should then continue to accept input (if possible).</p>
</blockquote>
<h2 id="sample-inputs--outputs">Sample Inputs &amp; Outputs</h2>
<p>Here&rsquo;s an example of an expected input for the program:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-tex" data-lang="tex"><span style="display:flex;"><span>5.0
</span></span><span style="display:flex;"><span>3
</span></span><span style="display:flex;"><span>8.5
</span></span><span style="display:flex;"><span>7
</span></span><span style="display:flex;"><span>k
</span></span><span style="display:flex;"><span>2.3.6
</span></span><span style="display:flex;"><span>0.0</span></span></code></pre></div><p>Here is the correct output for that input:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-tex" data-lang="tex"><span style="display:flex;"><span>ValueError: could not convert string to float: &#39;k&#39;
</span></span><span style="display:flex;"><span>ValueError: could not convert string to float: &#39;2.3.6&#39;
</span></span><span style="display:flex;"><span>Maximum: 8.5
</span></span><span style="display:flex;"><span>Minimum: 3.0</span></span></code></pre></div><p>Pause a moment to think about control flow, loops and exceptions you might need to handle.</p>
<p>When first starting out in coding and exception handling, it is often best to first write the code <strong>without</strong> exception handling, test it, then add the exception stuff.  Exception handlers can hide logic errors.</p>
<h2 id="handle-input">Handle Input</h2>
<p>First, we can write code to handle user input. The problem statement tells us to read inputs, one per line, until the user inputs 0. We also know that the user can input either whole numbers or floating point numbers. So, to make it simple, we&rsquo;ll just use floating point numbers for everything&ndash; this will require us to import math for number comparisons.  Now, math.isclose() cannot handle None, so instead we&rsquo;ll need to pick an initial value for our loop condition variable that will allow us to enter the loop, i.e. make the loop condition <code>True</code>.</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>x <span style="color:#f92672">=</span> <span style="color:#ae81ff">1.0</span>  <span style="color:#75715e"># math.isclose() cannot handle None, wo instead we pick an initial value for x </span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">while</span> <span style="color:#f92672">not</span> math<span style="color:#f92672">.</span>isclose(x,<span style="color:#ae81ff">0.0</span>):
</span></span><span style="display:flex;"><span>  string <span style="color:#f92672">=</span> reader<span style="color:#f92672">.</span>readline()
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">if</span> len(string) <span style="color:#f92672">==</span> <span style="color:#ae81ff">0</span>: 
</span></span><span style="display:flex;"><span>    x <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.0</span>
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>    x <span style="color:#f92672">=</span> float(string)</span></span></code></pre></div><p>If there is a a blank line, the <code>strip()</code> method will return the empty string.  Our program should take the the same action as if seeing a 0.0 so we assign 0.0 to x.  You should at this point test your logic , then add exception handlers.  Here we know that if <code>string</code> cannot convert until we&rsquo;ll get a ValueError.</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>x <span style="color:#f92672">=</span> <span style="color:#ae81ff">1.0</span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">while</span> <span style="color:#f92672">not</span> math<span style="color:#f92672">.</span>isclose(x, <span style="color:#ae81ff">0.0</span>):
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>        string <span style="color:#f92672">=</span> reader<span style="color:#f92672">.</span>readline()<span style="color:#f92672">.</span>strip()
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">if</span> len(string) <span style="color:#f92672">==</span> <span style="color:#ae81ff">0</span>:
</span></span><span style="display:flex;"><span>            x <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.0</span>
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>            x <span style="color:#f92672">=</span> float(string)
</span></span><span style="display:flex;"><span>            <span style="color:#66d9ef">if</span> <span style="color:#f92672">not</span> math<span style="color:#f92672">.</span>isclose(x, <span style="color:#ae81ff">0.0</span>):  
</span></span><span style="display:flex;"><span>                <span style="color:#66d9ef">if</span> x <span style="color:#f92672">&lt;</span> minimum <span style="color:#f92672">or</span> math<span style="color:#f92672">.</span>isclose(minimum, <span style="color:#ae81ff">0.0</span>):
</span></span><span style="display:flex;"><span>                    minimum <span style="color:#f92672">=</span> x
</span></span><span style="display:flex;"><span>                <span style="color:#66d9ef">if</span> x <span style="color:#f92672">&gt;</span> maximum <span style="color:#f92672">or</span> math<span style="color:#f92672">.</span>isclose(maximum, <span style="color:#ae81ff">0.0</span>):
</span></span><span style="display:flex;"><span>                    maximum <span style="color:#f92672">=</span> x
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>        print(<span style="color:#e6db74">&#34;ValueError: </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">&#34;</span><span style="color:#f92672">.</span>format(e))</span></span></code></pre></div><p>This should handle most exceptions that will occur in this program. Of course, there are many other ways that this code could be structured as well.</p>
<h2 id="logic">Logic</h2>
<p>Once we have the code to handle input, we can write the code to deal with logic. Keeping track of the minimum and maximum of a set of inputs is pretty simple:</p>
<div class="wrap-code highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>maximum <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.0</span>
</span></span><span style="display:flex;"><span>minimum <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.0</span>
</span></span><span style="display:flex;"><span>x <span style="color:#f92672">=</span> <span style="color:#ae81ff">1.0</span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">while</span> <span style="color:#f92672">not</span> math<span style="color:#f92672">.</span>isclose(x, <span style="color:#ae81ff">0.0</span>):
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>        string <span style="color:#f92672">=</span> reader<span style="color:#f92672">.</span>readline()<span style="color:#f92672">.</span>strip()
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">if</span> len(string) <span style="color:#f92672">==</span> <span style="color:#ae81ff">0</span>:
</span></span><span style="display:flex;"><span>            x <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.0</span>
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>            x <span style="color:#f92672">=</span> float(string)
</span></span><span style="display:flex;"><span>            <span style="color:#66d9ef">if</span> <span style="color:#f92672">not</span> math<span style="color:#f92672">.</span>isclose(x, <span style="color:#ae81ff">0.0</span>):  
</span></span><span style="display:flex;"><span>                <span style="color:#66d9ef">if</span> x <span style="color:#f92672">&lt;</span> minimum <span style="color:#f92672">or</span> math<span style="color:#f92672">.</span>isclose(minimum, <span style="color:#ae81ff">0.0</span>):
</span></span><span style="display:flex;"><span>                    minimum <span style="color:#f92672">=</span> x
</span></span><span style="display:flex;"><span>                <span style="color:#66d9ef">if</span> x <span style="color:#f92672">&gt;</span> maximum <span style="color:#f92672">or</span> math<span style="color:#f92672">.</span>isclose(maximum, <span style="color:#ae81ff">0.0</span>):
</span></span><span style="display:flex;"><span>                    maximum <span style="color:#f92672">=</span> x
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span>        print(<span style="color:#e6db74">&#34;ValueError: </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">&#34;</span><span style="color:#f92672">.</span>format(e))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>    
</span></span><span style="display:flex;"><span>print(<span style="color:#e6db74">&#34;Maximum: </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">&#34;</span><span style="color:#f92672">.</span>format(maximum))
</span></span><span style="display:flex;"><span>print(<span style="color:#e6db74">&#34;Minimum: </span><span style="color:#e6db74">{}</span><span style="color:#e6db74">&#34;</span><span style="color:#f92672">.</span>format(minimum))</span></span></code></pre></div><p>We can simply create variables to track the minimum and maximum values. Since our program won&rsquo;t accept $0$ as an input, we can use that as the default value for these variables.  Our logic does not use any functions likely to throw new errors, so no more exception handling is required.</p>
<h2 id="testing">Testing</h2>
<p>Finally, we may want to do some quick testing of our program. In theory, it should handle any inputs provided, including files that don&rsquo;t exist as command-line arguments and more. When testing, we should always see if we can find some input that breaks our program, then adjust the program as needed to prevent that problem. Of course, we should always make sure that it produces the correct answers, too.</p>

  
    
    
  
  
<div class="box notices cstyle noiframe">
  <div class="box-label"><i class="fa-fw fas fa-times-circle"></i> Web Only</div>
  <div class="box-content">

<p>This content is presented in the course directly through Codio. Any references to interactive portions are only relevant for that interface. This content is included here as reference only.</p>
</div>
</div>

            <footer class="footline">

            </footer>
          </article>

          <section>
            <h1 class="a11y-only">Subsections of A Worked Example</h1>
    
    
          </section>
          </section>

        </div>
      </main>
    
<div class="git-footer">
<p class="theme-version-footer">5.18.0</p>
<p>Last modified by: 
            <i class='fas fa-user'></i> Russell Feldhausen
            <i class='fas fa-calendar'></i> <a href="https://gitlab.cs.ksu.edu/cc/210/generic/textbook/-/commit/dc1990296b86b4f7f39165f7648d744727faa7f9">Oct 25, 2023</a>
</p>
</div>

    
    </div>
    <script src="https://ksu-cs-textbooks.github.io/cc210/js/clipboard.min.js?1701361006" defer></script>
    <script src="https://ksu-cs-textbooks.github.io/cc210/js/perfect-scrollbar.min.js?1701361006" defer></script>
    <script src="https://ksu-cs-textbooks.github.io/cc210/js/theme.js?1701361006" defer></script>
  </body>
</html>
